nuxeo-features/rest-api/nuxeo-rest-api-server/src/main/java/org/nuxeo/ecm/restapi/server/jaxrs/BatchUploadObject.java

```diff
@@ -139,96 +138,110 @@ protected Response uploadNoTransaction(@Context HttpServletRequest request,
+                        uploadChunkIndex, fileSize);
             }
         } else {
-            fileEntry = bm.getFileEntry(batchId, fileIdx);
-            if (fileEntry == null) {
-                return buildEmptyResponse(Status.NOT_FOUND);
+            if (fileName != null) {
+                fileName = URLDecoder.decode(fileName, "UTF-8");
+            }
+            try (InputStream is = request.getInputStream()) {
+                addStream(uploadType, batchId, fileIdx, is, fileName, mimeType, uploadedSize, chunkCount,
+                        uploadChunkIndex, fileSize);
             }
-        }
-
-        if (fileEntry == null && UPLOAD_TYPE_CHUNKED.equals(uploadType)) {
-            fileEntry = bm.getFileEntry(batchId, fileIdx);
         }
 
         StatusType status = Status.CREATED;
         Map<String, Object> result = new HashMap<>();
         result.put("uploaded", "true");
         result.put("batchId", batchId);
         result.put("fileIdx", fileIdx);
-        result.put("uploadedSize", uploadedSize);
-        if (fileEntry != null && fileEntry.isChunked()) {
-            result.put("uploadType", UPLOAD_TYPE_CHUNKED);
-            result.put("uploadedChunkIds", fileEntry.getOrderedChunkIndexes());
-            result.put("chunkCount", fileEntry.getChunkCount());
-            if (!fileEntry.isChunksCompleted()) {
-                status = new ResumeIncompleteStatusType();
+        result.put("uploadType", uploadType);
+        result.put("uploadedSize", String.valueOf(uploadedSize));
+        if (UPLOAD_TYPE_CHUNKED.equals(uploadType)) {
+            BatchFileEntry fileEntry = Framework.getService(BatchManager.class).getFileEntry(batchId, fileIdx);
+            if (fileEntry != null) {
+                result.put("uploadedChunkIds", fileEntry.getOrderedChunkIndexes());
+                result.put("chunkCount", fileEntry.getChunkCount());
+                if (!fileEntry.isChunksCompleted()) {
+                    status = new ResumeIncompleteStatusType();
+                }
             }
-        } else {
-            result.put("uploadType", UPLOAD_TYPE_NORMAL);
         }
         return buildResponse(status, result, isMultipart);
     }
 
+    protected long getUploadedSize(HttpServletRequest request) {
+        String contentLength = request.getHeader("Content-Length");
+        if (contentLength == null) {
+            return -1;
+        }
+        return Long.parseLong(contentLength);
+    }
+
+    protected void addStream(String uploadType, String batchId, String fileIdx, InputStream is, String fileName,
+            String mimeType, long uploadedSize, int chunkCount, int uploadChunkIndex, long fileSize)
+            throws IOException {
+        String uploadedSizeDisplay = uploadedSize > -1 ? uploadedSize + "b" : "unknown size";
+        if (UPLOAD_TYPE_CHUNKED.equals(uploadType)) {
+            log.debug(String.format("Uploading chunk [index=%d / total=%d] (%s) for file %s", uploadChunkIndex,
```

Comment:
I'm wondering here if this code is called a lot or not, construction of log message could hit down performance. Maybe we can add a `isDebugEnabled()`, wdyt ?